lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart

```diff
@@ -44,12 +44,28 @@ abstract class ViewRasterizer {
   /// Draws the [layerTree] to the screen for the view associated with this
   /// rasterizer.
   Future<void> draw(LayerTree layerTree) async {
-    final ui.Size frameSize = view.physicalSize;
+    ui.Size frameSize = view.physicalSize;
     if (frameSize.isEmpty) {
       // Available drawing area is empty. Skip drawing.
       return;
     }
 
+    // The [frameSize] may be slightly imprecise if the `devicePixelRatio` isn't
+    // an integer. For example, is you zoom to 110% in Chrome on a Macbook, the
+    // `devicePixelRatio` is `2.200000047683716`, so when the physical size is
+    // computed by multiplying the logical size by the devie pixel ratio, the
+    // result is slightly imprecise as well. Nevertheless, the number should
+    // be close to an integer, so round the frame size to be more precice.
+    bool isCloseToIntegerSize(ui.Size size) {
+      return (size.width - size.width.roundToDouble()).abs() < 0.001 &&
+          (size.height - size.height.roundToDouble()).abs() < 0.001;
+    }
+
+    assert(isCloseToIntegerSize(frameSize),
+        'Physical size is in fractional pixels.');
```

Comment:
This seems to be asserting a condition that's not under our control. `view.physicalSize` seems to be supplied by the browser. What does it mean for the assert not to pass? Would it indicate a bug in the web engine? Or a bug in the app code?

Reply:
`view.physicalSize` is the result of multiplying `window.visualViewport.{width,height}` by `window.devicePixelRatio` where `window.visualViewport.{width,height}` is the width and height of the screen in CSS pixels and `window.devicePixelRatio` is the ratio of CSS pixels to physical pixels.

I actually can't find anywhere that says the physical pixels must be a whole number. So I changed this to check _if_ the computed physical size is very close to an integer, then just round it.

Reply:
Ok, a condition makes sense. I think you explained this to me in person but I forgot, why do we not want to use `ceilToDouble()`?

Another question, if the size of the frame changes due to rounding but the CSS size of the canvas stays the same, is there risk of reintroducing https://github.com/flutter/flutter/issues/122541?

Reply:
The render canvas that we use to display the scene is sized to exactly the size of the window, and the bitmap we transfer into it from the OffscreenCanvas is stretched to fit inside the render canvas. To avoid issues like https://github.com/flutter/flutter/issues/122541, we must ensure the size of the bitmap we put into the render canvas is exactly the same as the size of the canvas in physical pixels. Due to issues outlined in the PR comments, computing the physical size is inexact by a tiny amount, so neither `floor()` nor `ceil()` are appropriate to massage the computed physical size values to a whole number of pixels (because the computed physical size can be either slightly below or slightly above the actual physical size), which is why `round()` is used.

So, for example, if you used `ceilToDouble()` and the `physicalSize` you get from the FlutterView is 100.00001 x 100.00001, then you will create a bitmap that is 101 x 101 pixels, and when you transfer that to the render canvas which is sized exactly to be 100px x 100px, then you will get blurriness issues.